kernel/timeout.c

```diff
@@ -195,6 +158,42 @@ void z_set_timeout_expiry(s32_t ticks, bool idle)
 	}
 }
 
+void z_clock_announce(s32_t ticks)
+{
+#ifdef CONFIG_TIMESLICING
+	z_time_slice(ticks);
+#endif
+
+	k_spinlock_key_t key = k_spin_lock(&timeout_lock);
+
+	announce_remaining = ticks;
+
+	while (first() != NULL && first()->dticks <= announce_remaining) {
+		struct _timeout *t = first();
+		int dt = t->dticks;
+
+		curr_tick += dt;
```

Comment:
Looking at this compared with the previous `z_clock_announce` I'm unable to detect any difference in observable behavior.  The commit message describes the change as:
> The z_clock_announce() implementation had a bug when multiple timeouts
    were expiring simultaneousl.  The idea of "current time" would be
    adjusted after the first one, so identical timeouts set by the first
    and subsequent ones would get different expiration times

If `curr_tick` meas the "current time", the rewrite and the original *appear* to be doing the same thing: advancing it by the delay required for the about-to-expire timer, so each expiration callback has `curr_tick` equal to the time at which the timer was expected to fire.

What am I missing?

Reply:
The bug was (I think...) that newly added timeouts from underneath the t->fn() callback invocation would be pre-decremented by the remaining announce time. But now I'm not seeing that in the old code either and need to go back to figure out exactly what I was seeing. The new code is simpler at least. :)

Reply:
Addressed in new commit message.